source/Variant.cpp

```diff
@@ -0,0 +1,279 @@
+			}
+			else
+			{
+				// If given the name of a ship, remove all ships by that name from this variant.
+				string shipName = child.Token(1);
+				auto ShipToRemove = [&](const Ship *s) noexcept -> bool
+				{
+					return s->VariantName() == shipName;
+				};
+
+				auto removeIt = remove_if(ships.begin(), ships.end(), ShipToRemove);
+				if(removeIt != ships.end())
+					ships.erase(removeIt, ships.end());
+				else
+					child.PrintTrace("Warning: Did not find matching ship for specified operation:");
+			}
+		}
+		else
+		{
+			// If this is a subsequent call of Load(), clear the variant
+			// if we aren't adding to it.
+			if(reset && !add)
+			{
+				reset = false;
+				variants.clear();
+				ships.clear();
+			}
+
+			int n = 1;
+			int index = 1 + add;
+			if(variant)
+			{
+				string variantName;
+				if(child.Size() >= index + 1 && !child.IsNumber(index))
+				{
+					variantName = child.Token(index++);
+					if(variantName == name)
+					{
+						node.PrintTrace("Warning: A variant cannot reference itself:");
+						continue;
+					}
+				}
+
+				if(child.Size() >= index + 1 && child.Value(index) >= 1.)
+					n = child.Value(index);
+
+				// If this variant is named, then look for it in GameData.
+				// Otherwise this is a new variant definition only for this variant.
+				if(!variantName.empty())
+				{
+					variants.emplace_back(GameData::Variants().Get(variantName), n);
+					if(child.HasChildren())
+						child.PrintTrace("Warning: Skipping children of named variant in fleet definition:");
+				}
+				else
+					variants.emplace_back(Variant(child, add), n);
+			}
+			else
+			{
+				if(child.Size() >= index + 1 && child.Value(index) >= 1.)
+					n = child.Value(index);
+				ships.insert(ships.end(), n, GameData::Ships().Get(child.Token(index - 1)));
+			}
+		}
+	}
+}
+
+
+
+void Variant::FinishLoading()
+{
+	// Prevent a variant from containing itself.
+	if(!name.empty())
+		for(auto it = variants.begin(); it != variants.end(); )
+		{
+			if(it->Get().NestedInSelf(name))
+			{
+				it = variants.eraseAt(it);
+				Files::LogError("Error: Infinite loop detected and removed in variant \"" + name + "\".");
+			}
+			else
+				++it;
+		}
+}
+
+
+
+bool Variant::IsValid() const
+{
+	// A variant should have at least one valid ship or nested variant.
+	bool validShips = any_of(ships.begin(), ships.end(),
+			[](const Ship *const s) noexcept -> bool { return s->IsValid(); });
+	bool validVariants = any_of(variants.begin(), variants.end(),
+			[](const WeightedVariant &v) noexcept -> bool { return v.Get().IsValid(); });
+
+	return (validShips || validVariants);
+}
+
+
+
+// Choose a list of ships from this variant. All ships from the ships
+// vector are chosen, as well as a random selection of ships from any
+// nested variants in the WeightedList.
+vector<const Ship *> Variant::ChooseShips() const
+{
+	vector<const Ship *> chosenShips = ships;
```

Comment:
So somehow things are crashing on this line, or about here. If I put a Files::LogError above and below this, sometimes I only get the line above this printing.
```
# No ships in this variant?
Fleet::Place test fleet
1 # Number of variants to choose from
	Variant::ChooseShips
	0 # Number of chosenShips
	0 # Number of nested variants
	0 # Number of ships returned
0 # Number of ships returned
# Then, with only one variant to choose from, there's somehow a ship now.
Fleet::Place test fleet
1
	Variant::ChooseShips
	1
	0
	1
1
# Then back to no ships?
Fleet::Place test fleet
1
	Variant::ChooseShips
	0
	0
	0
0
Fleet::Place test fleet
1
	Variant::ChooseShips
# Crashes
```

Reply:
It's accessing trash data. The variant being removed has a weight of 3 and the remaining variant has a weight of 1, yet after the removal of the variant with a nested variant within it, the number of variants left correctly drops to 1 but the total weight is only changing to 3. No idea why, though.

Reply:
So either there's something wrong with the range erase of WeightedList or it has something to do with using remove_if, because I switched back to using the older syntax (see below) and it works fine and properly adjusts the weight.
```c++
bool didRemove = false;
Variant toRemove(child, true);
for(auto it = variants.begin(); it != variants.end(); ++it)
	if(it->Get() == toRemove)
	{
		it = variants.eraseAt(it);
		didRemove = true;
	}

if(!didRemove)
	child.PrintTrace("Warning: Did not find matching variant for specified operation:");
```

Reply:
Alright, it'd definitely the use of the remove_if at fault somehow (and likely something that WeightedList should be doing that it isn't).
If I order the variants like this it crashes:
```
fleet "test fleet"
	government "Merchant"
	names "civilian"
	variant 3
		variant "test variant"
	variant 1
		Flivver
```
But if I order them like this it's fine and works as expected:
```
fleet "test fleet"
	government "Merchant"
	names "civilian"
	variant 1
		Flivver
	variant 3
		variant "test variant"
```
To my understanding remove_if is suppose to push the objects to be removed to the end of the list, but the list doesn't seem like it's actually getting modified, yet it's still erasing the last object in the list since it knows that one object is suppose to be erased.